File: 3754-maximum-manhattan-distance-after-k-changes/3754-maximum-manhattan-distance-after-k-changes.go
package main


func helper(A, B [2]int, k int) int {
	mia := min(A[0], k)
	A[0] -= mia
	k -= mia
	B[1] += mia

	mib := min(B[0], k)
	B[0] -= mib
	k -= mib
	A[1] += mib

	return A[1] - A[0] + B[1] - B[0]
}

func helper2(ns, ew [2]int, k int) int {
	if ns[0] > ns[1] {
		ns[0], ns[1] = ns[1], ns[0]
	}
	if ew[0] > ew[1] {
		ew[0], ew[1] = ew[1], ew[0]
	}
	return max(helper(ns, ew, k), helper(ew, ns, k))
}

func maxDistance(s string, k int) int {
	ns := [2]int{0, 0}
	ew := [2]int{0, 0}
	res := 0

	for _, ch := range s {
		if ch == 'N' {
			ns[0]++
		}
		if ch == 'S' {
			ns[1]++
		}
		if ch == 'E' {
			ew[0]++
		}
		if ch == 'W' {
			ew[1]++
		}
		res = max(res, helper2(ns, ew, k))
	}

	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: 3754-maximum-manhattan-distance-after-k-changes/3754-maximum-manhattan-distance-after-k-changes.cpp

class Solution {
    int helper(pair<int,int> A, pair<int, int> B, int k) {
        if(A.first > A.second) swap(A.first, A.second);
        if(B.first > B.second) swap(B.first, B.second);
        int mia = min(A.first, k);
        A.first -= mia, k -= mia;
        B.second += mia;
        int mib = min(B.first, k);
        B.first -= mib, k -= mib;
        A.second += mib;
        return A.second - A.first + B.second - B.first;
    }
public:
    int maxDistance(string s, int k) {
        pair<int,int> ns{0,0}, ew{0,0};
        int res = 0;
        for(auto& ch : s) {
            if(ch == 'N') ns.first++;
            if(ch == 'S') ns.second++;
            if(ch == 'E') ew.first++;
            if(ch == 'W') ew.second++;
            res = max(res, helper(ns,ew,k));
        }
        return res;
    }
};